### DISCUSSION TITLE
uneven task completion times in EM step -- suboptimal tuning?

### DISCUSSION THREAD

**Discussion by ericmanning** [moj-analytical-services/splink#254] [Q&A]
Thanks again for the brilliant package. I am currently using splink with Python 3.7 and Spark 2.4 to deduplicate (publicly-available) campaign contribution data in the U.S. This package has worked very well on data from smaller states (14k rows), completing the task in a few minutes as expected.

On (relatively) larger states, ~ 200k rows, regardless of the number of dataframe partitions or specified `spark.sql.shuffle.partitions` or `spark.default.parallelism`, I am unable to overcome uneven task completion times in the EM step.

For example, for anywhere between a few dozen or a few hundred `spark.sql.shuffle.partitions`, the last few tasks take far longer than the others, resulting in both suboptimal CPU usage and long runtimes (see below, x-axis is time of day in HH:MM -- this last task will drag on for ~ 30 minutes). This issue is far worse when weighting one or more fields by term frequency.

<img width="509" alt="Screen Shot 2021-12-10 at 2 06 45 AM" src="https://user-images.githubusercontent.com/43050489/145532867-25dbdb72-dd26-4528-9555-e8d34883397f.png">

Things I have tried:
- More (or fewer) data partitions (from 15 to 1000) [The default behavior reading in my .parquet file is only one partition, so I repartition, specifying no key]
- Changing `spark.sql.shuffle.partitions` and `spark.default.parallelism`
- Replacing nulls in data with empty strings (join problems?)

Exact code (both the slurm script for my cluster job and the python script) -- as well as a log file for a cluster job that was terminated early (the last few tasks will drag on) -- are attached, if it's any help.

I would be grateful for any guidance the authors could provide for Spark parameter tuning, given that I'm not sure how tasks are being allocated in the package. I do suspect that this is is a tuning problem, but I'm not sure what else to try.

[job.txt](https://github.com/moj-analytical-services/splink/files/7690834/job.txt)
[dedupe_state.txt](https://github.com/moj-analytical-services/splink/files/7690835/dedupe_state.txt)
[slurm-39028725_3.txt](https://github.com/moj-analytical-services/splink/files/7690847/slurm-39028725_3.txt)

**Comment by RobinL** [comment]
Hiya.  Thanks for the info, and especially providing the code.

This issue is probably with your blocking rules:

```
"blocking_rules": [
        "(l.contbr_city = r.contbr_city) AND ((l.po_num is null AND r.po_num is null) OR (l.po_num is not null AND r.po_num is not null)) AND ((l.gender = r.gender) OR (l.gender = 'm' AND r.gender = 'u') OR (l.gender = 'u' AND r.gender = 'm') OR (l.gender = 'f' AND r.gender = 'u') OR (l.gender = 'u' AND r.gender = 'f'))",
        "(l.contbr_nm_last = r.contbr_nm_last) AND ((l.po_num is null AND r.po_num is null) OR (l.po_num is not null AND r.po_num is not null)) AND ((l.gender = r.gender) OR (l.gender = 'm' AND r.gender = 'u') OR (l.gender = 'u' AND r.gender = 'm') OR (l.gender = 'f' AND r.gender = 'u') OR (l.gender = 'u' AND r.gender = 'f'))"
    ],
 ```

In a nutshell, you can't have OR conditions in blocking rules for large jobs.  This is a result of how Spark performs joins.

Blocking rules are implemented by performing an `inner join` of the input dataset on itself.

For example, if a blocking rule is 'l.surname = r.surname', then Splink is doing something like:

```
select *
from df as l
INNER JOIN df AS r 
ON l.surname = r.surname
```

Spark performs the inner join using hash partitioning on the equality condition, which enables effective parallelisation.

Unfortunately, Spark is unable to optimise  (effectively parallelise) OR conditions in an inner join.



If you want to achieve an OR condition, you're better to have multiple blocking rules. 

So, for example, if you wanted to achieve `l.surname = r.surname OR l.first_name = r.first_name`, you'd be better with two blocking rules:
```
blocking_rules = ['l.surname = r.surname', 'l.first_name = r.first_name'].  
```
Both rules now have an equality condition, and it'll run faster

I'm going to move this to a discussion, since at the moment I don't think it's a bug.   Our docs could obviously be better, but finding time for comprehensive docs is an ongoing struggle .

If, afte rfurther investigation, we find a bug, we can turn it back into an issue

**Comment by ericmanning** [comment]
Thank you for responding to this! Unfortunately, revising the blocking rules did not quite solve the issue I'm having.

I did, however, remove the blocking rules altogether (only 220k rows in the dataframe) and CPU usage improved and yields expected behavior. (At a certain point, though, with larger states this will run out of memory.)

This leads me to believe that the bottleneck is somewhere in the join processes, the sizes of which seem determined not by the number of data partitions, but the size of the blocks. Because I'm working with heavily skewed data (in Arizona, for example, most people live in one of a small number of cities) a few of these joins are much larger than the rest.

Given this working hypothesis, I'm curious whether the code (either in the package or Spark's default behavior) salts the join keys? If I have correctly identified the issue, salting would hypothetically fix the bottleneck.

Thanks again for being so responsive to questions.

**Comment by RobinL** [comment]
Thanks.  I think you're right that the bottleneck is in the join process.  We do not implement salting in the join process, because we've never figured out a way of doing this in a performant way (although we do have some ideas).  The problem is that it's not the 'classic' case  of salting that you see most often - the inner join in question is generating record comparisons (creating a kind of 'cartesian product within each block' - so for example, if name is  John Smith, we need to create the cartesian product of record comparisons amongst records where full name = John Smith.  

The problem, as you say, is with skew.  Let's say you blocked on first name and surname - i.e a blocking rule of `l.first_name = r.first_name and l.surname = r.surname`.   All 'John Smith' records are sent to a single executor, which then creates the cartesian product.  So if there are 100 'John Smiths' you get (100 * 99)/2 = 4,950 record comparisons.  That would be viable, but if there were 10,000 Smiths, you have a problem.  

The salting problem is then finding a way of:
- Splitting up the 'John Smiths' so some are sent to different executors
- But ensuring you create ALL 'john smith to john smith' comparisons, NOT just the comparisons within the 'salted block'

The closest we've got is adding a salt, let's say an random int with a value between 1 and 10 and then blocking on something like.
 `l.first_name = r.first_name and l.surname = r.surname and l.salt = 1`
`l.first_name = r.first_name and l.surname = r.surname and l.salt = 2`
...
`l.first_name = r.first_name and l.surname = r.surname and l.salt = 10`

But we haven't found a way of implementing this in the general case in a way that's both simple and performant.

You can learn a bit more about this by turning on logging at the debug level.  In the `splink_demos` tutorials, you can do this with:

```
import logging

logging.basicConfig()  # Means logs will print in Jupyter Lab

# Set to DEBUG if you want splink to log the SQL statements it's executing under the hood
logging.getLogger("splink").setLevel(logging.DEBUG)
```

This will then log all the SQL statements generated by Splink, which will give you a bit of a better idea of how blocking works


The good news is that I think there are other ways round your problem which should fix it

**Comment by RobinL** [comment]
As a followup to this, I've tried to articulate the salting problem as clearly as possible on Stackoverflow.  [Here](https://stackoverflow.com/questions/70317018/pyspark-salting-an-inner-join-in-the-presence-of-skew) is the question I've posted.

**Comment by ericmanning** [comment]
Thanks for this! I don't know much about Spark itself beyond what I've learned to implement this package, but would an upgrade to Spark 3.x.x solve this without any code changes ([adaptive query execution on joins](https://spark.apache.org/docs/latest/sql-performance-tuning.html#optimizing-skew-join))?

**Comment by RobinL** [comment]
I'd totally forgotten about that. I did some testing a couple of months ago on it, and it does help. You're right to point to it, because I think it probably is the answer to the Stackoverflow question (d'oh!).  The performance gain is relatively modest - something like 30%.

Basically there are three things that affect the performance of the blocking rules:
1). How loose they are.  This determines the total number of record comparisons Spark will generate.  It is dictated by the equality conditions in the blocking rules, and the cardinality of the fields.
2). Whether there's anything in the blocking rules that 'confuses' Spark into failing to optimise the inner join.  `OR` in blocking rules is a particular 'gotya' here to be avoided.
3). The skew of the blocking rules.  Whilst blocking on (say), first and last name may be fine for most names, a large group of 'John Smiths' could cause problems.  Whilst AQE may help a bit here, it doesn't solve this problem because fundamentally, if there are 10,000 John Smiths, that's (10,000*9,999/2)=50m record comparisons that have to be generated amongst John Smiths.

**Comment by ericmanning** [comment]
If, as you note in the StackOverflow comment, ADE assigns tasks proportional to the size of the blocks (rather than ~ blocks^2, as one would need), then ADE is going to help less (relative to optimal assignment) as the size of my blocks increase. My data is highly skewed by city/postal_code, so the solution here for me is probably the one you outlined above: manually salt. I'll probably assign salt values randomly **within** the city/postal_code blocks, with N salt values proportional to (size of block)^2, and see what happens.

A fundamental issue here is that the data I'm using is _so_ messy that it's difficult to find reasonable blocking rules that don't drop obvious matches with some regularity. I've played around with clustering strings using affinity propagation on a matrix of first name string similarity, e.g., to reduce the size of the largest blocks, but these all blow up on large states as well. Might just have to cut my losses.
